feat: config hot-reload

sammcj opened the pull request:
This PR adds a new feature:

- Hot-reload when the config file changes.

It is disabled by default and as per the additions to the readme can be enabled by running llama-swap with the `--watch-config` flag.

The behaviour is such that it will wait for any in-flight requests to complete before reloading the underlying llama server.

_Also added some very minor readme linting fixes while I was in there 😅_

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Added live configuration reload capability, allowing the server to automatically reload its configuration file without downtime when changes are detected.
	- Introduced a new command-line flag to enable or disable automatic config file watching and reloading.

- **Improvements**
	- Enhanced server lifecycle management for more graceful startup, shutdown, and process handling.
	- Logging output is now less verbose and more informative, especially during normal startup, shutdown, and reload events.
	- Documentation updated with clearer instructions, improved code block formatting, and detailed descriptions of available command-line flags.

- **Bug Fixes**
	- Reduced risk of port collisions in tests by assigning explicit ports and adding delays where necessary.

- **Tests**
	- Updated tests to reflect changes in handler method naming and improved test diagnostics.
	- Adjusted test process commands for clearer output during test runs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mostlygeek reviewed (changes requested):
Overall this looks like it's in the right direction. I think the right design is for all reloading due to configuration changes be outside of ProxyManager. It is instantiated with a `New(config *Config)` and I want to keep the contract that the config comes in from the outside and is never modified inside.

sammcj commented:
Thanks for the review @mostlygeek, 

I've updated the PR to move all config reload logic out of ProxyManager.

Key changes:

1. ProxyManager now maintains its contract: it's instantiated with a config and never modifies it internally
2. Config watching/reloading happens entirely in the main package
3. When config changes:
	- Main package creates a new ProxyManager instance
	- Handles graceful shutdown of the old instance
	- Starts the new instance with updated config

This maintains better separation of concerns while preserving all functionality. The changes also made the reload process more robust by properly coordinating server lifecycle.

Let me know if you'd like me to explain any part in more detail or if there's any further changes you'd like made.

mostlygeek commented:
It's looking good. I'm just pushed a big PR #109 which changes llama-swap.go a bit. I think there may be some small merge conflicts. Since ProxyManager's contract hasn't changed that part will be fine :). 

Right now `ProxyManager.Run()` is used to create a listener. What do you think of changing that a bit so we can replace ProxyManager without dropping the TCP connection?

Maybe something like: 

```golang
  proxyManager := proxy.New(config)
  srv := &http.Server{
    Addr:    ":8080",
    Handler: proxyManager,
  }

  go func() {
    // service connections
    if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
      log.Fatalf("listen: %s\n", err)
    }
  }() 
```

ProxyManager is an [http.Handler](https://pkg.go.dev/net/http#Handler). Perhaps it's possible to change srv.Handler gracefully? I haven't looked too much into this.

sammcj commented:
Yeah that would be nice actually, good thinking.

What I'll do is update this branch first, then once you've merged your PR, I'll pull and fix conflicts etc...

I'll mark it as draft until that's done :)

sammcj commented:
OK that was a bit more challenging than I anticipated and pushed the levels of my go skills so I'm sorry if I've overcomplicated anything but this is now working well on my server @mostlygeek

mostlygeek commented:
That took me a little bit to review. Most of it looks pretty good. I'm not entirely sure what was going on with test changes but I left comments. 

BTW: if this is [you](https://www.reddit.com/r/LocalLLaMA/comments/1fwj055/comment/lqezrcl/), thanks for following through 😄 !

mostlygeek commented:
I think this would be a good change to ProxyManager: 

```golang
func NewWithLoggers(config Config, muxLogger *LogMonitor, proxyLogger *LogMonitor, upstreamLogger *LogMonitor) *ProxyManager { 
   ... 
}
```

This way: 

- loggers are defined at the main() level
- loggers are shared across old and new proxy manager instances
- the log stream history is maintained across restarts
- the fmt.Println() in `llama-swap.go` can use the loggers directly

sammcj commented:
OK I've OK I've rebased from main now.

I did make a couple of changes to your logging here:

```
-		p.proxyLogger.Warnf("<%s> cmd or cmd.Process is nil", p.ID)
+		p.proxyLogger.Debugf("<%s> cmd or cmd.Process is nil (normal during config reload)", p.ID)
```

Just because with how the process handler works now, this would cause what looks to be an alarming message every time the config reloads, but I can revert those back if you wish @mostlygeek

sammcj commented:
Ah it seems tests are working locally but not in CI, I'll have to come back to this another day sorry.

mostlygeek commented:
> Ah it seems tests are working locally but not in CI, I'll have to come back to this another day sorry.

Thanks. I'm too tired tonight to review anyways :)

mostlygeek reviewed (changes requested):
I took a look at things and the changes in llama-swap.go look OK. I'm confused a bit by the changes in proxymanager.go and proxymanager_test.go. A few things: 

in llama-swap.go: 

- I think `reloadChan` could be `reloadChan := make(chan struct{})` to keep the shutting and replacement logic in the same place. 

In proxymanager_test.go: 

- revert TestProxyManager_Shutdown() to the previous one. The test changed a lot and it looks to me like it lost it's original purpose. 
- maybe a new new test to make sure rotation works? Though it should as most of the changes were external and there was just a bit of renaming / logging improvements. 
- most of the changes in this file should be for the renaming of HandlerFunc to ServeHTTP

in logmonitor_test.go: 

- what were the changes in logmonitor_test.go for? I don't see the purpose, perhaps consider reverting it to the original.

sammcj commented on proxy/logMonitor_test.go:
```diff
@@ -21,27 +21,39 @@ func TestLogMonitor(t *testing.T) {
 
 	var wg sync.WaitGroup
-	wg.Add(1)
+	wg.Add(2) // One for each client
```
This was made because the test was refactored to use two goroutines instead of one. Previously, a single goroutine handled message collection from two clients.

In the updated version:

1. One goroutine is now dedicated to collecting messages for `client1`.
2. A second, new goroutine is dedicated to collecting messages for `client2`.

Each of these two goroutines calls `wg.Done()` upon its completion.

sammcj commented:
rfm

sammcj commented on proxy/proxymanager.go:
```diff
@@ -223,22 +226,19 @@ func (pm *ProxyManager) stopProcesses() {
 	pm.Lock()
 	defer pm.Unlock()
 
-	// shutdown process in parallel
```
`ProxyManager.Shutdown()` should now correctly initiats the shutdown of `ProcessGroup`s in parallel using goroutines and a `sync.WaitGroup`. The `StopProcesses()` method also operates in parallel at the `ProcessGroup` level.

sammcj commented on proxy/proxymanager_test.go:
```diff
@@ -257,60 +257,110 @@ func TestProxyManager_ProfileNonMember(t *testing.T) {
 	}
 }
 
 func TestProxyManager_Shutdown(t *testing.T) {
```
Thanks, test reverted!

sammcj commented:
I _think_ I've addressed the remaining issues, but if my lack of go skills are causing you more grief than it's worth - I totally understand if you'd rather reject this and do it your own way.

By the way - that CodeRabbit bot is good!

mostlygeek commented on proxy/process.go:
```diff
@@ -410,7 +414,20 @@ func (p *Process) ProxyRequest(w http.ResponseWriter, r *http.Request) {
 	// prevent new requests from being made while stopping or irrecoverable
 	currentState := p.CurrentState()
-	if currentState == StateFailed || currentState == StateShutdown || currentState == StateStopping {
+
```
There is a chance here that the upstream os process is still running/resident in memory. Simply resetting the state to `StateStopped` to have `p.start()` is likely not enough. I think we'd need something that additionally makes sure it is no longer there with a SIGTERM, and with that doesn't work a SIGKILL. 

I have been considering a more robust attempted recovery behaviour for failed processes. That enhancement should be in a different PR not this one. 

Can you revert this behaviour change. We can also ensure that this PR stays more focused on the original intent.

sammcj replied:
Yeah I hit that issue when I was testing my PRs changes (hence my attempt to fix it), I've reverted this now to leave things as they were.

mostlygeek commented on proxy/process_test.go:
```diff
@@ -18,20 +18,21 @@ var (
 
```
I don't understand these changes to the tests. I would suggest reverting them along with the behaviour changes in process.go.

sammcj replied:
- Reverted process_test.go (I can no longer make tests pass locally without the change so am relying Github Actions feedback now).
- The only changes left in process.go are to reduce the info level logging noise when handling config hot-reloads where the processes are expected be interrupted and stopped.

I couldn't get the process tests to pass locally without making changes, I found the `TestProcess_AutomaticallyStartsUpstream` had issues with port allocation.

```
=== RUN   TestProcess_AutomaticallyStartsUpstream
    process_test.go:43:
        	Error Trace:	/Users/samm/git/llama-swap/proxy/process_test.go:43
        	Error:      	Not equal:
        	            	expected: "ready"
        	            	actual  : "failed"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(proxy.ProcessState) (len=5) "ready"
        	            	+(proxy.ProcessState) (len=6) "failed"

        	Test:       	TestProcess_AutomaticallyStartsUpstream
    process_test.go:45:
        	Error Trace:	/Users/samm/git/llama-swap/proxy/process_test.go:45
        	Error:      	Not equal:
        	            	expected: 200
        	            	actual  : 502
        	Test:       	TestProcess_AutomaticallyStartsUpstream
        	Messages:   	Expected status code 200, got 502
    process_test.go:46:
        	Error Trace:	/Users/samm/git/llama-swap/proxy/process_test.go:46
        	Error:      	"unable to start process: upstream command exited unexpectedly: exit status 1\n" does not contain "testing91931"
        	Test:       	TestProcess_AutomaticallyStartsUpstream
    process_test.go:59: Expected status code 200, got 503
--- FAIL: TestProcess_AutomaticallyStartsUpstream (0.25s)
```

---

As an aside - if it helps to explain my reasoning - I passed the PR diff with and without the process_test changes to a LLM to challenge my assumptions and got it to explain why the tests were failing without changes:

Q: "Do the changes in my PR _require_ this test to change?"

A: "No, your PR's core features (config reloading) do *not* fundamentally break the `Process` type's existing contract in a way that *necessitates* a change to the test's logic to make it pass if the test were perfectly stable.

However, to achieve __stable and reliable tests__ in the presence of your PR (and generally), `TestProcess_AutomaticallyStartsUpstream` (and other similar tests like `TestProcess_WaitOnMultipleStarts`) *do* benefit significantly from more robust port management (i.e., using distinct, hardcoded or carefully allocated ports for each `simple-responder` instance within the test). This makes the test less dependent on global state and timing.

You might need to explain to the reviewer that the reverted version of `TestProcess_AutomaticallyStartsUpstream` is inherently flaky due to its port allocation strategy, and the changes you had previously made (using `getTestSimpleResponderConfigPort` with distinct ports) were to ensure test reliability, not to work around a bug introduced by the config reloading feature. The PR might be making this flakiness more apparent, but the fix lies in making the test more robust."

mostlygeek reviewed (changes requested):
Hey @sammcj we're almost there. 

The main changes are reverting things that are not related, or have unclear rational to the original intent of the PR. The functional changes to process.go/process_test.go are a good examples. 

I am getting a bit of review fatigue but the finish line looks close. :)

mostlygeek reviewed (approved):
Looks good!

mostlygeek commented:
Thanks @sammcj! 🥳
